### PR TITLE
HARDFORK: Fix mining centralisation

### PR DESCRIPTION
**Pull request by luke-jr**
This solves mining centralisation by enabling GPU miners to be the leading-edge again. (Hopefully the next generation of ASIC miners will have learned their lesson, so we don't need to do this again.)

### CODE REVIEW

**Comment by NateBrune** [comment]
@luke-jr just fyi for more direct contact you can chat at bitcoinclassic.slack.com

**Comment by transcats** [comment]
This just redistributes mining for a few years. Memory-hard POW has better potential as a "fix".

In practice it's going to be virtually impossible to get consensus for this. Better off starting a whole new blockchain.

**Comment by luke-jr** [comment]
@transcats Memory-hard PoW hasn't been invented yet, so we can't implement that.

This is literally no harder to get consensus for than a hardforked block size increase. This is probably easier, in fact, since it impacts fewer people negatively.

**Comment by jl2012** [comment]
@transcats "This just redistributes mining for a few years." Just hardfork again. If we could do it the first time, we will have the second time

**Comment by sandakersmann** [comment]
Looks like the master troll has arrived...

**Comment by tonesnotes** [comment]
Luke is not a troll. He does think outside the box.
But in this case his suggestion is orthogonal to the purpose at hand.

**Comment by luke-jr** [comment]
If we're going to hardfork, it makes sense to fix this problem.

**Comment by judahmu** [comment]
Do you have the start of a risks document?  51% attacks at the time of switch over, anything like that?

**Comment by luke-jr** [comment]
The risks are the same as with any hardfork, including the already-planned 2 MB size change.

**Comment by tonesnotes** [comment]
There is an existential risk to Classic in not being extremely focused on making the first release a rapid success with existing miners.

**Comment by jcliff** [comment]
NACK. This renders hundreds of millions of dollars worth of equipment useless. It will leave a bad taste in the miners' mouth. I also think this is an unnecessary fix to a problem that won't exist in the long term. ASIC mining hardware will sooner or later hit fundamental physical limits and become commoditized. Decentralization of mining will naturally follow.

File: src/txdb.cpp [hunk 1]
```diff
@@ -203,7 +203,7 @@ bool CBlockTreeDB::LoadBlockIndexGuts()
                 pindexNew->nStatus        = diskindex.nStatus;
                 pindexNew->nTx            = diskindex.nTx;
 
-                if (!CheckProofOfWork(pindexNew->GetBlockHash(), pindexNew->nBits, Params().GetConsensus()))
+                if (!CheckProofOfWork(pindexNew->GetBlockHash(), pindexNew->nBits, pindexNew->nTime, Params().GetConsensus()))
```
Comment (orangeman) [inline src/txdb.cpp:206]: This is actually a really good idea!
Mining centralisation could be mitigated this way :+1: 

Full ACK!

**Comment by jtoomim** [comment]
NACK.

**Comment by mshean** [comment]
^ That wasn't very democratic

**Comment by orangeman** [comment]
Please explain _why_ you closed this pull request?
Do you have any reasonable reasons?

**Comment by pennyfx** [comment]
This is a good idea.  It's like creating a new Litecoin.

**Comment by jtoomim** [comment]
@mshean if you think we should change the PoW function, going straight to Github is _not_ the correct method. This is better discussed on Consider.it or reddit or something like that.

In the mean time, I think that this thread is getting trolled, and I recommend that people chill out for a few days while tempers calm down, and we should talk things over.

**Comment by orangeman** [comment]
Isn't this an open source project?

**Comment by jameshilliard** [comment]
@jtoomim Reddit is pretty terrible due to brigading and vote bots, consider.it isn't much better.

**Comment by luke-jr** [comment]
We were just having a discussion on this on Slack. You closed it without even discussing?! :(

Reopen it, this is something to seriously consider.

**Comment by jtoomim** [comment]
@luke-jr discuss radical ideas in other forums first, please.

**Comment by pennyfx** [comment]
@luke-jr   This also ensures bitcoin-classic is an altcoin being bootstrapped from nothing.

**Comment by orangeman** [comment]
Is there an alternative plan to fix centralisation?

**Comment by jameshilliard** [comment]
@jtoomim I think communication is a big issue for this project, other forums generally aren't good places for developers to talk since there are too many non-technical users. Social media is not the right place for technical discussion.

**Comment by dgmib** [comment]
NACK

**Comment by jtoomim** [comment]
If you do not like the forums that this project uses, and do not like being downvoted, you are welcome to participate in a different project.

If you want to participate in the dialog here in a meaningful fashion, you are welcome to. That means learning how to use the tools that we use as best as possible.

If you really think that you have a democratic majority behind this PoW function, you are welcome to independently convince users to adopt it.

If you continue to troll, you will be ignored.

**Comment by pennyfx** [comment]
@orangeman The plan is to let the market do its thing.  SHA256 ASICs will be commoditized, it just takes time to play out.  No matter what you use for PoW, an ASIC will be created if the market price is high enough.

**Comment by jtoomim** [comment]
@pennyfx don't feed the trolls, please.

**Comment by luke-jr** [comment]
This is not any more radical or altcoin-y than the hardfork to 2 MB.

**Comment by pennyfx** [comment]
@luke-jr So accidently hardforks are ok, but planned hardforks are radical?

**Comment by jtoomim** [comment]
@pennyfx don't feed the trolls, please.

**Comment by orangeman** [comment]
Why are you not allowing a debate?
This is very discouraging. :-1:

**Comment by jameshilliard** [comment]
@jtoomim Are you trying to force developers to go to particular forums such as reddit where it is easier to bury discussions?

**Comment by orangeman** [comment]
Let's be constructive!

How do you intend to fix this instead?

The problem will not go away by doing nothing..

**Comment by justusranvier** [comment]
@orangeman If you want to be constructive:
1. This PR supposes that a problem exists, called "mining centralization"
2. Describe the mining centralization in detail, giving specific criteria that can be used to compare two theoretical situations and distinguish between one that has mining centralization, and one that does not. Without a null hypothesis, it's not possible to say whether or not the posited condition even exists at all.
3. Describe why the posited problem requires a solution, i.e. in what negative effective outcomes will Bitcoin users avoid by solving the problem.
4. Describe the costs of the solution.

Once it's been established that a problem exists which should be fixed, then it would be time to look at all potential solutions.

Talking about a specific solution before the problem has even been proven to exist or be a problem at all is premature.

**Comment by jtoomim** [comment]
@justusranvier Don't feed the trolls, please. They can have the last comment, it's fine.

**Comment by justusranvier** [comment]
@jtoomim One way or another, those questions need to be answered.

Of course the trolls aren't going to make any good faith attempts to do so, and that's why it's important to point out what they could be doing instead of trolling.

**Comment by toomim** [comment]
I don't believe in trolls. I don't believe in NACK. Luke is welcome to submit this proposal to bitcoin classic. Is there a problem with submitting a pull request in github before proposing it in text? I don't see one. Who cares which comes first?

**Comment by hannemennah** [comment]
@luke-jr: if you think this is a good idea, why have you not submitted a similar pull request to the Bitcoin Core repository?

@jtoomim, I completely understand where you're coming from, but please do remember to be careful about shutting things down mid-debate; it's a slippery slope.

**Comment by toomim** [comment]
@hannemennah: I welcome all Bitcoin Core developers to submit pull requests to the Classic repository instead of the Core repository.

Also, I agree that we don't want to shut down the debate.

However, @everyone: Bitcoin Classic will not release anything but the 2MB hardfork patch until we have hard forked. We are focused on the hard fork. That's how I interpret the words I wrote on the website: https://bitcoinclassic.com. After we fork, we open the floodgates to new features as determined by the community.

If you disagree with this policy, please make a considerit proposal for a better policy and build support for it. **Edit:** eh, I should probably just make a considerit proposal for this policy myself! **Edit:** [done](https://bitcoinclassic.consider.it/first-release-is-2mb-hard-fork).

**Comment by surg0r** [comment]
Core devs finally woken up. Instead of improving Core they try and disrupt Classic with this distracting crap in a desperate ploy to dissuade miners from making the switch. 

Good luck.

NACK.

**Comment by luke-jr** [comment]
@hannemennah It has been discussed before, but at this time Core's approach does not involve a hardfork, and the PoW change cannot be done as a softfork (even if it could, softforks are in miners' hands and they obviously wouldn't go for it).

I do think Bitfury should be given the opportunity to make good on their public statement they will sell their new miners widely (and maybe other manufacturers will step forward and do the same), but there's a few months before the hardfork activates to allow for that, and we should begin the discussion now.

**Comment by Kirvx** [comment]
The current Bitcoin mining network represents hundreds millions dollars of investment, and this is what strongly determines the price of a Bitcoin.
If you change that, Bitcoin Classic will be finally an altcoin with a value approaching 0.
We already have tons of altcoins offering this kind of short term decentralization.

Apparently the goal of this pull request is to disqualify Bitcoin Classic so that Bitcoin Core have no competitors with other ideas.

@luke-jr You should convince others Core devs to fork for 2MB block before Bitcoin Core loses a part of its power and breaks the community a bit more.

**Comment by luke-jr** [comment]
There is no evidence mining costs have any influence on price; the history of Bitcoin suggests the opposite is true.

There's no need to convince Core devs or any other devs to do a hardfork; we don't have a choice in the matter. The final decision will be made by the economy, and developers must go along with that.

**Comment by speedmax** [comment]
@luke-jr Seriously, Do you have better things to do? 

You guys are just embarrassing yourself, in a big way.
Sad to see hackers doing child play.

**Comment by dskloet** [comment]
@speedmax don't feed the trolls.

**Comment by cryptonaut420** [comment]
This is a clear sabotage attempt as a direct result of many miners expressing support for Classic. Changing the PoW algorithm would be a massive stab in the back to miners and kill any chance of a hard fork happening. Luke knows this perfectly well.

**Comment by dskloet** [comment]
@cryptonaut420 it is indeed clear. No need to discuss it further :).

**Comment by mr-e-** [comment]
Thank you luke for your help in destroying core!

**Comment by p3yot3** [comment]
The trolling of luke-jr here only makes me want to use classic even more & rid myself of his nonsense once & for all. I hope everyone here remembers his comments when he is looking for a job.

**Comment by BadLibertarian** [comment]
@jtoomim I understand why this has been closed here, but why is it no longer showing up on consider.it? Maybe @luke-jr is trolling and maybe he isn't, but it's hard for me to make up my mind about that unless he can flesh out the idea. The fact that he didn't bother to do that before posting is not a point in his favor, but I agree with @hannemennah that this is a slippery slope.

Edit: my bad - it is still on consider.it. For some reason I didn't see it the first time I looked. Carry on.

**Comment by jtoomim** [comment]
@BadLibertarian I see it here: https://bitcoinclassic.consider.it/merge-6-fix-mining-centralisation?results=true

Currently, among all users, it has been voted on 36 times, and the mean vote is -0.62 (i.e. 19% support).

Among verified users, it has been voted on 12 times, and the mean vote is -0.79 (i.e. 11% support).

**Comment by ghost** [comment]
AMD is going to love this , reopen this if you are serious about community involvement

**Comment by paulmadore** [comment]
I firmly support this idea and I'm disappointed to see it didn't get much traction. I think Skein-1024 might have been a wiser option, but no matter. Moving away from the algorithm that so many large, centralized miners require would definitely bring Bitcoin back into the hands of the people. For a time.

**Comment by sturles** [comment]
Untested ACK from me.  If we are going to hardfork, it makes sense to fix as many problems as possible in this fork.  It could be improved by setting a life time for this hashing algorithm, and then change every year or so.

**Comment by ghost** [comment]
I think that the idea of decentralizing mining is more complicated than just a change in hash function. change the algo, and eventually someone will manage to create an asic , and someone with serious money will buy enough of those to control a significant portion of the hash rate. 

the main reason i can see a algo change making sense is going for a greener method.

**Comment by orangeman** [comment]
:1234: It seems we are not finding consense any time soon. But this is an important issue that needs more thorough discussion. Do we have any scientific data to proof that this would not work? Otherwise we could just merge and move on...

**Comment by jtoomim** [comment]
screen shot 2016-01-16 at 4 06 21 pm

**Comment by dskloet** [comment]
@jtoomim follow your own advice and don't feed trolls.

**Comment by orangeman** [comment]
What happened to the _agree_ votes?

**Comment by jtoomim** [comment]
screen shot 2016-01-16 at 4 15 08 pm

**Comment by alp-bitcoin** [comment]
ACK

**Comment by ghost** [comment]
well, look at whose votes counted , let the rest of us vote if you want to be taken seriously

**Comment by Zorlin** [comment]
NACK.

This is a freaking awful idea.

**Comment by Alex-Linhares** [comment]
LOL Nack.  Luke, take your pills.

**Comment by Aquentus** [comment]
LOL NACK. I thought his github account was hacked but he confirmed it wasn't!

**Comment by orangeman** [comment]
Github is technical discussion (peer review) about code. You are making this place like reddit! How do you think we should get any meaningful work done if everything gets debated over and over and over again? This project has become incapable of making any decision at all https://github.com/bitcoinclassic/bitcoinclassic/pull/14 If you continue like this the community will have to move on to a different project...

**Comment by priiduneemre** [comment]
IMHO this is an interesting idea, albeit very controversial (_i.e._ would probably cost BC a lot of starting momentum from miners). Thus probably not a very good fit for the first release, OTOH would sit well with the inevitable HF. Should probably be discussed in more detail via other venues first.

@Alex-Linhares: no personal attacks please. Keep it constructive.
@orangeman: it is also not your personal blog. Please cool off and be objective or take your tongue-in-cheek trolling elsewhere.

**Comment by chrisfranko** [comment]
I lold, lets do it.

**Comment by darkhosis** [comment]
jtoomim, that was awesome.

reminded me of;

https://bitcointalk.org/index.php?topic=90513.40

scroll down a bit to see my post (zvs) and lukejr's response. gg

**Comment by luke-jr** [comment]
More relevant links would be:
- https://bitcointalk.org/?topic=102040.msg1122034#msg1122034
- https://bitcointalk.org/?topic=99497.msg1154765#msg1154765

This is not a new idea. It has been planned for this specific situation since before mining ASICs existed.

Disclaimer: I made some incorrect statements in those links (we have all learned much over the years). For example, 50% is _not_ required to perform attacks.

**Comment by jtoomim** [comment]
I have unsubscribed from this thread. We will not be changing the PoW function in Classic.

**Comment by paulmadore** [comment]
Is someone here trying to imply that the hashing algorithm has to be Sha-256? That shows a severe lack of education, to say the least.

**Comment by darkhosis** [comment]
Wasn't talking about new ideas.  Rather what I said in a thread 3 1/2 years ago, that ASICs would 'obviously' increase centralized mining.   The only thing going against at first was 1) many of the initial companies needed startup funds, so some small percentage of people got ASICs at a good price, 2) all the idiots buying USBs for $80/each and w/e else.  Why would you sell something that's more profitable to keep?

re: your comment, 'Perhaps after eliminating all the illegal botnets and miners who are just mining for "free money" with no real interest, there might just be 2% of total miners now - but that's still better than the status quo. Total hash power has always been the important factor of Bitcoin mining, and it was designed to be intentionally highly competitive (and therefore just barely profitable).'

So, yes, we have a nice 180 here.  That's all, & no need to hear some other excuse, so I'll go ahead and unsubscribe as well, laugh.

**Comment by nomnombtc** [comment]
NACK this would crash bitcoins market value and destroy confidence.

Also I think it was fine to close the pull because it is very very unlikely to get merged anyway at this point, if situation changes it can always be reopened.

**Comment by Pr0methean** [comment]
Why not make both SHA256 and Keccak hashes valid for PoW for the 3 years or so that an ASIC lasts? They could even have different difficulty parameters with the Keccak difficulty falling as the SHA256 difficulty rose (relative to the overall difficulty as currently calculated).

**Comment by cypherdoc** [comment]
what a troll, lukejr.

NACK

**Comment by nachik** [comment]
I agree with paulmadore, Skein-1024 preferred over Keccak but in any case, since classic is a hardfork anyway it would be silly to miss the opportunity to decentralise mining at least for a number of years, while the subsidy is still high.

**Comment by LongShao007** [comment]
mark

**Comment by earonesty** [comment]
Sure, this shouldn't be in the first hardfork.     But something _like this_ needs to happen.   I thought the "enforced p2pool" concept was better.

**Comment by rromanchuk** [comment]
For those of you not actually understanding what's going on here, luke is stylistically deadpanning this pull request for lulz. This sarcastic POW change is essentially telling us to go [redacted] ourselves and make an "altcoin" It's actually pretty funny, but guys, it's a joke. 

Or maybe you are all brilliant and just meta-trolling me now.

**Comment by paulmadore** [comment]
Romanchuk, you seem to not understand what's going on here. This is not the first time this idea has been discussed, nor is Luke-Jr trolling, and you're out of line.

**Comment by orangeman** [comment]
Might be needed in case the _perceived_ economic majority is different from the _real_ economic majority.

**Comment by zptec** [comment]
Great idea.

**Comment by kristovatlas** [comment]
NACK

**Comment by mrzeta** [comment]
Great idea. Merge！！！！！！

**Comment by adityasuseno** [comment]
NACK
LoL
Just do that in Core

No need to do it for now. Unless there is proven serious weakness in SHA256. It may cause chaos among the miners. Also there is no such ASIC resistant algorithm.

**Comment by adityasuseno** [comment]
NACK
LoL
Please do that in Core

No need to do it for now. Unless there is proven serious weakness in SHA256. It may cause chaos among the miners. Also there is no such ASIC resistant algorithm.

**Comment by ripper234** [comment]
NACK lol indeed.

POW will only change to reply to an actual threat of the network e.g. Quantum Computers.

**Comment by rebroad** [comment]
I think it's certainly a good idea to decentralize mining if possible, but I would have thought it needs to be done gradually, not suddenly. From the code changes, it looks pretty sudden. Yes, such a change would require a hard-fork. I think if the power was gradually taken from ASICs and moved towards GPUs then even mining pools would be happy with this as it wouldn't mean they suddenly lose customers, although it might be hard for people to determine at which point to switch from ASICs to GPUs, I'm always in the favour of making something gradual. I'm actually against the sudden 1MB to 2MB block size change for the same reason. I believe this should be more gradual too, ideally allowing the miners to vote per block, and letting it increase (or decrease) by a small percentage per block (but I'm going off-topic potentially now).

**Pull request closed without merging**